File: cpp/703.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

class KthLargest {
public:
  int size;
  priority_queue<int> heap;
  KthLargest(int k, vector<int> &nums) {
    size = k;
    for (int i : nums) {
      heap.push(0 - i);
      if (heap.size() > size)
        heap.pop();
    }
  }

  int add(int val) {
    heap.push(0 - val);
    if (heap.size() > size)
        heap.pop();
    return 0 - heap.top();
  }
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */

File: go/703.go
package main

import (
	"container/heap"
)

type Heap []int

// Len implements heap.Interface
func (h Heap) Len() int {
	return len(h)
}

// Less implements heap.Interface
func (h Heap) Less(i int, j int) bool {
	return h[i] < h[j]
}

// Swap implements heap.Interface
func (h Heap) Swap(i int, j int) {
	h[i], h[j] = h[j], h[i]
}

// Pop implements heap.Interface
func (h *Heap) Pop() interface{} {
	ret := (*h)[len(*h)-1]
	*h = (*h)[:len(*h)-1]
	return ret
}

// Push implements heap.Interface
func (h *Heap) Push(x interface{}) {
	*h = append(*h, x.(int))
}

func (h *Heap) Peek() interface{} {
	return (*h)[0]
}

type KthLargest struct {
	heap *Heap
	size int
}

func Constructor(k int, nums []int) KthLargest {
	h := &Heap{}
	heap.Init(h)
	for _, v := range nums {
		heap.Push(h, v)
	}
	for h.Len() > k {
		heap.Pop(h)
	}
	return KthLargest{
		heap: h,
		size: k,
	}
}

func (this *KthLargest) Add(val int) int {
	heap.Push(this.heap, val)
	if this.heap.Len() > this.size {
		heap.Pop(this.heap)
	}
	return this.heap.Peek().(int)
}

/**
 * Your KthLargest object will be instantiated and called as such:
 * obj := Constructor(k, nums);
 * param_1 := obj.Add(val);
 */
